torch/jit/_trace.py

```diff
@@ -928,16 +934,27 @@ def register_submods(mod, prefix):
         module = make_module(mod, _module_class, _compilation_unit)
 
         for method_name, example_inputs in inputs.items():
-            # this is needed since Module.__call__ sets up some extra tracing
-            func = mod if method_name == "forward" else getattr(mod, method_name)
+            if method_name == "forward":
+                # "forward" is a special case because we need to trace
+                # `Module.__call__`, which sets up some extra tracing, but uses
+                # argument names of the real `Module.forward` method.
+                func = mod
+                forward_method = getattr(mod, method_name)
+                argument_names = get_callable_argument_names(forward_method)
+            else:
+                func = getattr(mod, method_name)
+                argument_names = get_callable_argument_names(func)
+
             example_inputs = make_tuple(example_inputs)
+
             module._c._create_method_from_trace(
                 method_name,
                 func,
                 example_inputs,
                 var_lookup_fn,
                 strict,
                 _force_outplace,
+                argument_names,
```

Comment:
Why not get these argument names in C++ by using `py::module(...)...`? Is it because you need to handle `forward` in a special way?

Reply:
Yeah, logic to handle module's `forward` is complicated. It would be very ugly in c++